web/packages/teleport/src/User/UserContext.tsx

```diff
@@ -52,20 +63,52 @@ export function useUser(): UserContextValue {
   return useContext(UserContext);
 }
 
+// ClusterPreferences represents a map of clusterId to their cluster preferences
+type ClusterPreferences = Record<string, UserClusterPreferences>;
+
 export function UserContextProvider(props: PropsWithChildren<unknown>) {
   const { attempt, run } = useAttempt('processing');
+  const clusterPreferences = useRef<ClusterPreferences>({});
 
   const [preferences, setPreferences] = useState<UserPreferences>(
     makeDefaultUserPreferences()
   );
 
+  const getClusterPinnedResources = useCallback(async (clusterId: string) => {
```

Comment:
We have to use `useCallback` here because it is used in `getPinnedResources` which in turn is used in `useEffect`.

Theoretically, we don't have to use `useCallback` in `updateClusterPinnedResources` (and `updatePinnedResources` attempt) because that function is only invoked on click (so no risk of infinite re-renders). 
@avatus you can get rid of it if you want.